### PR TITLE
expanded weights: group norm faster rule

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70141
* __->__ #73696
* #73694

Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

### CODE REVIEW

**Comment by samdow** [comment]
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719388).

**Comment by samdow** [comment]
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719388).

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2294,6 +2294,8 @@
 
 - func: group_norm(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enabled=True) -> Tensor
 
+- func: _group_norm_all_outputs(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05) -> (Tensor, Tensor, Tensor)
```
[the remark this answers is not in the document]
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:2297]: Yes, we don't expose forward/backward pairs to python.
`torch.nn.grad` was an attempt at that and as you can see from the current state. Never worked out as it is way too tricky.

For this one, why not just move the checks into the `native_group_norm` function (which should be private most likely btw).
They will be duplicated but I feel like this is better than having 3 functions for a single op.

File: aten/src/ATen/native/group_norm.cpp [hunk 2]
```diff
@@ -31,6 +31,29 @@ std::tuple<Tensor, Tensor, Tensor> native_group_norm(
...
+      " and input of shape ",
+      X.sizes());
+  TORCH_CHECK(
+      !beta.defined() || (beta.dim() == 1 && beta.numel() == C),
+      "Expected bias to be a vector of size equal to the number of ",
+      "channels in input, but got bias of shape ",
+      beta.sizes(),
+      " and input of shape ",
+      X.sizes());
+
```
Comment (zou3519) [inline aten/src/ATen/native/group_norm.cpp:56]: Is it possible to write a `check_group_norm_inputs(...)` function that does this and share it between native_group_norm and group_norm?

We should also leave a comment about why this shouldn't be deleted (someone will come along and inevitably try to delete the additional check). To defend against deletion we probably want to add a test somewhere for aten.ops.native_group_norm that asserts that it checks for these conditions.

File: torch/nn/utils/_expanded_weights/group_norm_expanded_weights.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    def forward(ctx, kwarg_names, _, *expanded_args_and_kwargs):
+        expanded_args, expanded_kwargs = standard_kwargs(kwarg_names, expanded_args_and_kwargs)
+        input, num_groups = expanded_args
+        N = input.shape[0]
+        C = input.shape[1]
+        HxW = reduce(operator.mul, input.shape[2:], 1)
+        weight, bias, eps = expanded_kwargs['weight'], expanded_kwargs['bias'], expanded_kwargs['eps']
+        output, mean, rstd = forward_helper(torch.native_group_norm, (input, weight, bias, N, C, HxW, num_groups, eps), {})
+        ctx.input, ctx.num_groups = input, num_groups
+        ctx.weight, ctx.bias, ctx.eps = weight, bias, eps
```
Comment (zou3519) [inline torch/nn/utils/_expanded_weights/group_norm_expanded_weights.py:22]: Optimization: we don't have to save weight, bias if none of them require gradients, right? (this is the case where someone defines nn.GroupNorm with affine=False.

Doesn't need to be handled in this PR, could be a followup. Also because I am not sure how we'd go about testing this (is there a test case for nn.GroupNorm with affine=False?)

**Review by zou3519 — APPROVED** [review]
I left a suggestion for refactoring the native_group_norm check so that it is shared between native_group_norm and group_norm and adding a test to defend against future people who will try to delete the check. Other than that, this LGTM

File: aten/src/ATen/native/group_norm.cpp (hunk 2 above)
Comment (samdow) [reply aten/src/ATen/native/group_norm.cpp]: Moved to helper function, added comment, also added test that expanded weights fails where it should in group norm--I figure that this should be sufficient since we don't exactly care if someone changes what happens in `native_group_norm` if they did it in a way that made expanded weights still correctly fail with incorrect arguments

File: torch/nn/utils/_expanded_weights/group_norm_expanded_weights.py (hunk 3 above)
Comment (samdow) [reply torch/nn/utils/_expanded_weights/group_norm_expanded_weights.py]: Added optimization (weight will still be needed for group norm backward, but able to optimize it out if neither input nor weight require gradients)! Group Norm does have an affine=False test but might want to double check when input + weight don't require grad but bias does...
